### PR TITLE
Sleeping if internet connection is disrupted (instead of erroring out)

### LINKED ISSUE #39: telliot stops when internet connection is lost
When the internet connection is lost, the telliot process stops running.

Expected behavior is that telliot would keep running until internet connection is reestablished then resume the process.

See attached for log with the traceback. (feel free to ping me in discord if attachment doesn't work) 
[log.txt](https://github.com/tellor-io/telliot-feed-examples/files/7620766/log.txt)

### PR DESCRIPTION
**Pull request by tallywiesenberg**
Closes #39 
This PR prevents Telliot from quitting with an Error if the internet connection is disrupted.

### Summary
- creates a function `is_online()` that pings cloudflare at `http://1.1.1.1` and returns True if ping was successful
- Implements this function in the `while True` loop of each reporter, logging a warning if internet is lost, then sleeping as usual

### Steps Taken to QA Changes
...

### Checklist

This pull request is:

- [ ] A documentation error, docs update, or typographical error fix
	- No tests or issue needed
- [x] A code fix
	- Please reference the related issue by including "Closes `<link to issue>`" in this Pull Request's summary section. 
        - If no issue exists, please create a bug report issue 
	- Please include tests. Fixes without tests will not be accepted unless it's related to the documentation only.
    - Please make sure docs are updated if need be
- [x] A feature implementation
	- Please reference the related issue by including "Closes `<link to issue>`" in this Pull Request's summary section. 
        - If no issue exists, please create a feature enhancement issue 
	- Please include tests
    - Please make sure docs updates are both thorough and easy to reproduce by somebody with limited knowledge of the feature that you are submitting

**Happy engineering!**

### CODE REVIEW

File: src/telliot_feeds/reporters/rng_interval.py [hunk 1]
```diff
@@ -311,6 +312,9 @@ async def report(self) -> None:
         """Submit latest values to the TellorFlex oracle."""
         logger.info(f"RNG reporting interval: {INTERVAL} seconds")
         while True:
+            online = await is_online()
+            if not online:
+                logger.warning("Unable to connect to the internet!")
             _, _ = await self.report_once()
```
Comment (oraclown) [inline src/telliot_feeds/reporters/rng_interval.py:318]: This one tries to report even if no internet

File: src/telliot_feeds/reporters/utils.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+"""Utils for Telliot Reporter classes"""
+import requests
+
+
+async def is_online() -> bool:
```
Comment (oraclown) [inline src/telliot_feeds/reporters/utils.py:5]: Add docstring

**Review by oraclown — CHANGES_REQUESTED** [review]
Thanks @tallywiesenberg ! Add tests for both expected behavior plz. Tests for your new function, and then a reporter trying to report when there's no internet, and also when there is (but mock the report_once func so it actually doesn't report)

File: src/telliot_feeds/cli/commands/report.py [hunk 3]
```diff
@@ -27,7 +27,7 @@
 logger = get_logger(__name__)
 
 
-TELLOR_FLEX_CHAINS = (137, 122, 80001, 3, 69, 1666600000, 1666700000, 421611)
+TELLOR_FLEX_CHAINS = (137, 122, 80001, 3, 69, 1666600000, 1666700000, 421611, 941)
```
[the remark this answers is not in the document]
Comment (tallywiesenberg) [reply src/telliot_feeds/cli/commands/report.py:30]: Ah I think i commited this to main by accident earlier. it adds pulsechain (chain id 941)

File: src/telliot_feeds/reporters/rng_interval.py (hunk 1 above)
Comment (tallywiesenberg) [reply src/telliot_feeds/reporters/rng_interval.py:318]: true, good catch

File: src/telliot_feeds/reporters/utils.py (hunk 2 above)
Comment (tallywiesenberg) [reply src/telliot_feeds/reporters/utils.py:5]: good catch thanks

**Comment by oraclown** [comment]
Possible solution for the testing the `report()` func updates:
```python
async def report(self) -> None:
        """Report values for pool reference assets & settle pools."""
        while True:
            online = await is_online()
            if online:
                _, _ = await self.report_once()
                _ = await self.settle_pools()
            else:
                logger.warning("Unable to connect to the internet!")

            logger.info(f"Sleeping for {self.wait_period} seconds")
            await asyncio.sleep(self.wait_period)
```
Use `mock.patch` to make the `asyncio.sleep()` call quit out of the while loop or exit

On another note, I think `if online` reads better than `if not online` and the logic flows better. Wbu?

**Comment by tallywiesenberg** [comment]
> Possible solution for the testing the `report()` func updates:
> 
> ```python
> async def report(self) -> None:
>         """Report values for pool reference assets & settle pools."""
>         while True:
>             online = await is_online()
>             if online:
>                 _, _ = await self.report_once()
>                 _ = await self.settle_pools()
>             else:
>                 logger.warning("Unable to connect to the internet!")
> 
>             logger.info(f"Sleeping for {self.wait_period} seconds")
>             await asyncio.sleep(self.wait_period)
> ```
> 
> Use `mock.patch` to make the `asyncio.sleep()` call quit out of the while loop or exit
> 
> On another note, I think `if online` reads better than `if not online` and the logic flows better. Wbu?

I concur, thank you for this

**Comment by oraclown** [comment]
per our call, an approach to fix the tests:
```python
# patch builtin while looop to just go once
    r = reporter
    r.check_is_online = lambda: return False
    r.wait_period = 0
    _ = await report once
    
    assert "blah" in caplog.text
 ```
 and move `is_onlin()` to general utils

**Comment by oraclown** [comment]
Looks like some conflicts to resolve first...

File: src/telliot_feeds/reporters/rng_interval.py [hunk 4]
```diff
@@ -313,8 +313,10 @@ async def report(self) -> None:
         logger.info(f"RNG reporting interval: {INTERVAL} seconds")
         while True:
             online = await is_online()
-            if not online:
+            if online:
+                _, _ = await self.report_once()
+                logger.info(f"Sleeping for {self.wait_period} seconds")
```
Comment (oraclown) [inline src/telliot_feeds/reporters/rng_interval.py:317]: move the sleeping msg out of the condition and put above where it sleeps. It's going to sleep whether it's online or not

File: src/telliot_feeds/reporters/tellorflex.py [hunk 5]
```diff
@@ -351,9 +351,10 @@ async def report(self) -> None:
 
         while True:
             online = await is_online()
-            if not online:
-                logger.warning("Unable to connect to the internet!")
-            else:
+            print("online status: ", online)
```
Comment (oraclown) [inline src/telliot_feeds/reporters/tellorflex.py:366]: rmv print

File: tests/reporters/test_polygon_reporter.py [hunk 6]
```diff
@@ -121,3 +122,18 @@ async def _fetch_gas_price():
     assert not staked
     assert not status.ok
     assert "Unable to fetch matic gas price for staking" in status.error
+
+@pytest.mark.asyncio
+async def test_reporting_without_internet(polygon_reporter):
```
Comment (oraclown) [inline tests/reporters/test_polygon_reporter.py]: I think this test isn't finished

Comment (oraclown) [reply tests/reporters/test_polygon_reporter.py]: or needs to be cleaned up

**Review by oraclown — CHANGES_REQUESTED** [review]
Thanks @tallywiesenberg . Some fixes needed

**Review by oraclown — APPROVED** [review]
Thanks @tallywiesenberg !

**Pull request merged**